writeups/picomini-2021-diana/notepad.md

```diff
@@ -0,0 +1,101 @@
+  </h3>
+  {% include "errors/" + error + ".html" ignore missing %}
+{% endif %}
+```
+
+How is this useful? Well, we can create our own error template by using our path traversal technique to create a file in the `templates/errors` folder. This new template can then be rendered on the index page if we pass it in as a query parameter.
+
+We can test a sample payload now:
+```
+..\\templates\\errors\\
+```
+
+If we paste that into the text area on the index page and submit, we get redirected to a `Not Found` error page. Not to worry! Since we wrote to a file outside of the static folder, we can't access it from `/static/...`, which resulted in the 404. However, we can check out our payload by passing our filename (minus extension) into `error`. The path I got redirected to was `///templates//errors//-7-kh4_p7dTY.html` so I went to:
+```
+https://notepad.mars.picoctf.net/?error=-7-kh4_p7dTY
+```
+
+Lo and behold, our payload was printed back to us!
+
+# SSTI Injection
+Since the error page is included as a Jinja template and not as a static page, we can utilize a server-side template injection to retrieve the flag. I found a great [reference](https://medium.com/@nyomanpradipta120/ssti-in-flask-jinja2-20b068fdaeee) that detailed how to implement a SSTI injection with the Jinja template engine.
+
+After the path of the `errors` folder, I padded the payload with `a` characters so that none of the injection was stored in the filename. While this is optional, I found that it made the request much cleaner because the filename isn't url encoded.
+
+The next step is to find the `string` class. This is usually done with `''.__class__` but because of the filter on underscores, we can't insert `__class__` directly into the `content` parameter. We can, however, insert it into a different query parameter and reference it that way. We can then get the string class's [method resolution order](https://www.geeksforgeeks.org/method-resolution-order-in-python-inheritance/) using `mro()`.
+
+The second class in the `string` class's MRO is the `object` class so we can index to 1.
+```py
+[<class 'str'>, <class 'object'>]
+```
+
+If we call `object.__subclasses__()`, we can access all the classes that are available. Since underscores are filtered, we can access the function the same way we accessed `__class__`. Our full payload is now:
+```py
+p = "..\\templates\\errors\\" + "a" * 128 + "{{ ''[request.args.get('class')].mro()[1][request.args.get('subclasses')]() }}"
+```
+
+When we access the index page with the filename passed into `error`, we get an internal server error. This is because we didn't pass in the `class` or `subclasses` query parameters. After setting them to `__class__` and `__subclasses__`, respectively, [a whole bunch of stuff is rendered](https://notepad.mars.picoctf.net/?error=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-oiv7lSJ0KoY&class=__class__&subclasses=__subclasses__).
```

Comment:
I would remove this link, because the challenge server could restart at any time and break it

Reply:
maybe a picture or a short snippet?